kv/kvserver/loqrecovery: TestReplicaCollection failed

kv/kvserver/loqrecovery.TestReplicaCollection [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12726227?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12726227?buildTab=artifacts#/) on release-23.1 @ [85392ec3a02bf344ab55b8e8154b27f5ab3f9a45](https://github.com/cockroachdb/cockroach/commits/85392ec3a02bf344ab55b8e8154b27f5ab3f9a45):


```
=== RUN   TestReplicaCollection
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/95fc0dff57a5bf3a6a9cc9666d301bb7/logTestReplicaCollection1234827256
    test_log_scope.go:79: use -show-logs to present logs inline
    server_integration_test.go:84: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/kv/kvserver/loqrecovery_test/pkg/kv/kvserver/loqrecovery/server_integration_test.go:84
        	            				github.com/cockroachdb/cockroach/pkg/kv/kvserver/loqrecovery_test/pkg/kv/kvserver/loqrecovery/server_integration_test.go:108
        	Error:      	Received unexpected error:
        	            	rpc error: code = Unknown desc = initial connection heartbeat failed: grpc: client cluster ID "7eaf8e4f-0f56-4acf-aac8-6fc49424b447" doesn't match server cluster ID "318556da-fca7-4b8f-80ab-c014d76a7b16" [code 2/Unknown]
        	Test:       	TestReplicaCollection
        	Messages:   	failed to retrieve replica info
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/95fc0dff57a5bf3a6a9cc9666d301bb7/logTestReplicaCollection1234827256
--- FAIL: TestReplicaCollection (13.17s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/replication
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicaCollection.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>



Jira issue: CRDB-33609

--- a/pkg/kv/kvserver/loqrecovery/server_integration_test.go
+++ b/pkg/kv/kvserver/loqrecovery/server_integration_test.go
@@ -52,6 +52,12 @@ func TestReplicaCollection(t *testing.T) {
 
 	ctx := context.Background()
 
+	// This test stops cluster servers. Use "reusable" listeners, otherwise the
+	// ports can be reused by other test clusters, and we may accidentally connect
+	// to a wrong node.
+	// TODO(pav-kv): force all tests calling StopServer to use sticky listeners.
+	listenerReg := listenerutil.NewListenerRegistry()
+	defer listenerReg.Close()
 	tc := testcluster.NewTestCluster(t, 3, base.TestClusterArgs{
 		ServerArgs: base.TestServerArgs{
 			StoreSpecs: []base.StoreSpec{{InMemory: true}},
@@ -62,6 +68,7 @@ func TestReplicaCollection(t *testing.T) {
 				},
 			},
 		},
+		ReusableListenerReg: listenerReg,
 	})
 	tc.Start(t)
 	defer tc.Stopper().Stop(ctx)

